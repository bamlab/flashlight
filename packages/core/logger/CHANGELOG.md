# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.1](https://github.com/bamlab/android-performance-profiler/compare/@perf-profiler/logger@0.3.0...@perf-profiler/logger@0.3.1) (2023-06-28)

**Note:** Version bump only for package @perf-profiler/logger

# 0.3.0 (2023-04-18)

### Bug Fixes

- **profiler:** improve timings accuracy ([#84](https://github.com/bamlab/android-performance-profiler/issues/84)) ([1908689](https://github.com/bamlab/android-performance-profiler/commit/19086891b618382dd290431e63cf72059a729133))

### BREAKING CHANGES

- **profiler:** if using a custom script, pollPerformanceMeasures will now take `{ onMeasure: (Measure) => void }` as second parameter

# 0.1.0-alpha.9 (2022-08-29)

### Features

- **logger:** add timestamps to log ([6928d3f](https://github.com/bamlab/android-performance-profiler/commit/6928d3fe5ec82544ab2e121b0f1ec2b70e2fcb94))

# 0.1.0-alpha.8 (2022-08-27)

### Features

- **profiler:** add spawn function ([e6eba2e](https://github.com/bamlab/android-performance-profiler/commit/e6eba2e88d9621fd096fc3f3b56614b46b9ca781))

# 0.1.0-alpha.7 (2022-08-12)

### Features

- **flipper:** prepare new version ([#9](https://github.com/bamlab/android-performance-profiler/issues/9)) ([4d97516](https://github.com/bamlab/android-performance-profiler/commit/4d97516f9a0b8f1715c0b22c1bdab70fb32cc527))

# 0.1.0-alpha.5 (2022-08-12)

# 0.1.0-alpha.3 (2022-08-04)

### Bug Fixes

- fix flipper plugin ([6035938](https://github.com/bamlab/android-performance-profiler/commit/6035938f8f0bcad14a32498babbf6a0ffacea607))
