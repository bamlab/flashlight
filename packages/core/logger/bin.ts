#!/usr/bin/env node

import { printExampleMessages } from "./index";

printExampleMessages();
