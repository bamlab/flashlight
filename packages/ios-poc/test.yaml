appId: org.reactjs.native.example.fakeStore
---
- tapOn: "VIENS ON PETE TOUT"
