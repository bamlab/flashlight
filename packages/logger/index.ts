import kleur from "kleur";
import { DateTime } from "luxon";

const info = kleur.blue;
const success = kleur.bold().green;
const warn = kleur.bold().yellow().bgRed;
const error = kleur.bold().red;
const timestampColor = kleur.grey;

export enum LogLevel {
  ERROR,
  WARN,
  SUCCESS,
  INFO,
  DEBUG,
}

let logLevel = LogLevel.INFO;

const log = (message: string) => {
  const timestamp = DateTime.now().toLocaleString(
    DateTime.TIME_24_WITH_SECONDS
  );
  const timestampLog = timestampColor(`[${timestamp}]`);
  console.log(`${timestampLog} ${message}`);
};

export const Logger = {
  setLogLevel: (level: LogLevel) => {
    logLevel = level;
  },
  debug: (message: string) => {
    if (logLevel < LogLevel.DEBUG) return;

    const time = performance.now();
    log(`🚧  ${Math.floor(time)}: ${message}`);
  },
  info: (message: string) => {
    if (logLevel < LogLevel.INFO) return;

    log(info(`ℹ️  ${message}`));
  },
  success: (message: string) => {
    if (logLevel < LogLevel.SUCCESS) return;

    log(success(`✅  ${message}`));
  },
  warn: (message: string) => {
    if (logLevel < LogLevel.WARN) return;

    log(warn(`⚠️  ${message}`));
  },
  error: (message: string) => {
    if (logLevel < LogLevel.ERROR) return;

    log(error(`🚨  ${message}`));
  },
};

export const printExampleMessages = () => {
  Logger.setLogLevel(Infinity);
  Object.keys(Logger).forEach((key: string) => {
    if (key === "setLogLevel") return;

    (Logger[key as keyof typeof Logger] as typeof Logger.debug)(
      `This is an awesome ${key} message`
    );
  });
};
