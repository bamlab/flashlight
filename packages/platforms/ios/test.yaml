appId: org.reactjs.native.example.fakeStore
---
- tapOn: "VIENS ON PETE TOUT"
- assertVisible: "ON A TOUT PETE"
