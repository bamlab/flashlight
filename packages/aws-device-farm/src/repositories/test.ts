import {
  ArtifactCategory,
  ArtifactType,
  ExecutionResult,
  ExecutionStatus,
  GetRunCommand,
  ListArtifactsCommand,
  ListRunsCommand,
  ScheduleRunCommand,
  TestType,
} from "@aws-sdk/client-device-farm";
import { execSync } from "child_process";
import { Logger } from "@perf-profiler/logger";
import { BaseRepository } from "./BaseRepository";

export class TestRepository extends BaseRepository {
  async getByName({ projectArn, name }: { projectArn: string; name: string }) {
    const { runs } = await this.client.send(
      new ListRunsCommand({ arn: projectArn })
    );

    return runs?.find((run) => run.name === name);
  }

  /**
   * aws devicefarm schedule-run --project-arn $PROJECT_ARN
   *  --app-arn $APK_UPLOAD_ARN
   *  --device-pool-arn $DEVICE_POOL_ARN
   *  --name Flash
   *  --test testSpecArn=$TEST_SPEC_UPLOAD_ARN,type=APPIUM_NODE,testPackageArn=$TESTS_UPLOAD_ARN
   */
  async scheduleRun({
    projectArn,
    apkUploadArn,
    devicePoolArn,
    testName,
    testPackageArn,
    testSpecArn,
  }: {
    projectArn: string;
    apkUploadArn: string;
    devicePoolArn: string;
    testName: string;
    testPackageArn: string;
    testSpecArn: string;
  }) {
    const { run } = await this.client.send(
      new ScheduleRunCommand({
        projectArn,
        appArn: apkUploadArn,
        devicePoolArn,
        name: testName,
        test: {
          type: TestType.APPIUM_NODE,
          testPackageArn,
          testSpecArn,
        },
      })
    );

    const testRunArn = run?.arn;

    if (!testRunArn) throw new Error(`Test run has no arn`);

    const consoleUrl = await this.getUrl({
      testRunArn,
    });
    // Hack to add the variable into Github actions although it shouldn't be done here
    if (process.env.GITHUB_ENV) {
      execSync(`echo "AWS_CONSOLE_TEST_RUN_URL=${consoleUrl}" >> $GITHUB_ENV`);
    }
    Logger.success(
      `Launched test ${testName}...
Follow along on ${consoleUrl}`
    );

    return testRunArn;
  }

  async isTestCompleted({ arn }: { arn: string }) {
    const { run } = await this.client.send(new GetRunCommand({ arn }));

    if (!run) throw new Error("Test run couldn't be found");

    Logger.info(`Test status is ${run.status}, result is ${run.result}`);

    if (run.status === ExecutionStatus.COMPLETED) {
      if (run.result !== ExecutionResult.PASSED) {
        throw new Error(`Test run has not succeeded, result is ${run.result}`);
      }

      return true;
    }

    return false;
  }

  async waitForCompletion({ arn }: { arn: string }) {
    while (!(await this.isTestCompleted({ arn }))) {
      Logger.info(`Waiting 10s for test to be done...`);
      await new Promise((resolve) => setTimeout(resolve, 10000));
    }
  }

  async getArtifactUrl({ arn, type }: { arn: string; type: ArtifactType }) {
    const { artifacts } = await this.client.send(
      new ListArtifactsCommand({
        arn,
        type: ArtifactCategory.FILE,
      })
    );

    const url = artifacts?.find((artifact) => artifact.type === type)?.url;

    if (!url) throw new Error(`Url not found for artifact type ${type}`);

    return url;
  }

  async getUrl({ testRunArn }: { testRunArn: string }) {
    const [projectId, runId] = testRunArn.split("run:")[1].split("/");
    const region = await this.client.config.region();

    return `https://${region}.console.aws.amazon.com/devicefarm/home?region=${region}#/mobile/projects/${projectId}/runs/${runId}/jobs/00000`;
  }
}
