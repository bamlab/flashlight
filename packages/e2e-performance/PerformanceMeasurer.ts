import { Logger } from "@perf-profiler/logger";
import {
  getPidId,
  Measure,
  pollPerformanceMeasures,
} from "@perf-profiler/profiler";
import { Trace } from "./Trace";

const waitFor = async <T>(
  evaluateResult: () => T | undefined | null,
  { timeout, checkInterval }: { timeout: number; checkInterval: number } = {
    timeout: 10000,
    checkInterval: 50,
  }
): Promise<T> => {
  if (timeout < 0) {
    throw new Error("Waited for condition which never happened");
  }
  const result = evaluateResult();
  if (result) return result;

  await new Promise((resolve) => setTimeout(resolve, checkInterval));

  return waitFor(evaluateResult, {
    timeout: timeout - checkInterval,
    checkInterval,
  });
};

export class PerformanceMeasurer {
  cpuMeasures: Measure[] = [];
  polling?: { stop: () => void };
  bundleId: string;
  shouldStop = false;
  timingTrace?: Trace;

  constructor(bundleId: string) {
    this.bundleId = bundleId;
  }

  async start() {
    const pid = await waitFor(
      () => {
        try {
          return getPidId(this.bundleId);
        } catch (error) {
          Logger.debug(`${this.bundleId} not yet started, retrying...`);
          return null;
        }
      },
      {
        timeout: 10000,
        // we don't add any timeout since `adb pidof` already takes a bit of time
        checkInterval: 0,
      }
    );

    this.timingTrace = new Trace();

    this.polling = pollPerformanceMeasures(pid, (measure) => {
      if (this.shouldStop) {
        this.polling?.stop();
      }

      this.cpuMeasures.push(measure);
      Logger.debug(`Received measure ${this.cpuMeasures.length}`);
    });
  }

  async stop(duration?: number) {
    const time = this.timingTrace?.stop();

    const TIME_INTERVAL_IN_MS = 500;
    if (duration) {
      // Hack to wait for the duration to be reached in case test case has finished before
      await waitFor(
        () => this.cpuMeasures.length * TIME_INTERVAL_IN_MS > duration,
        {
          checkInterval: TIME_INTERVAL_IN_MS,
          timeout: duration * 1000,
        }
      );
      this.cpuMeasures = this.cpuMeasures.slice(
        0,
        duration / TIME_INTERVAL_IN_MS + 1
      );
    } else {
      this.shouldStop = true;
      // Hack to wait for the last measures to be received
      await new Promise((resolve) =>
        setTimeout(resolve, TIME_INTERVAL_IN_MS * 2)
      );
    }

    // Ensure polling has stopped
    this.polling?.stop();

    return {
      time: time ?? 0,
      measures: this.cpuMeasures,
    };
  }
}
