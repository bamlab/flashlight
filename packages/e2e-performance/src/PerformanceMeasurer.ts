import { Logger } from "@perf-profiler/logger";
import {
  getPidId,
  Measure,
  pollPerformanceMeasures,
} from "@perf-profiler/profiler";
import { Trace } from "./Trace";
import { waitFor } from "./utils/waitFor";

export class PerformanceMeasurer {
  measures: Measure[] = [];
  polling?: { stop: () => void };
  bundleId: string;
  shouldStop = false;
  timingTrace?: Trace;

  constructor(bundleId: string) {
    this.bundleId = bundleId;
  }

  async start(
    onMeasure: (measure: Measure) => void = () => {
      // noop by default
    }
  ) {
    const pid = await waitFor(
      () => {
        try {
          return getPidId(this.bundleId);
        } catch (error) {
          Logger.debug(`${this.bundleId} not yet started, retrying...`);
          return null;
        }
      },
      {
        timeout: 10000,
        // we don't add any timeout since `adb pidof` already takes a bit of time
        checkInterval: 0,
      }
    );

    this.timingTrace = new Trace();

    this.polling = pollPerformanceMeasures(pid, (measure) => {
      if (this.shouldStop) {
        this.polling?.stop();
      }

      this.measures.push(measure);
      onMeasure(measure);
      Logger.debug(`Received measure ${this.measures.length}`);
    });
  }

  async stop(duration?: number) {
    const time = this.timingTrace?.stop();

    const TIME_INTERVAL_IN_MS = 500;
    if (duration) {
      // Hack to wait for the duration to be reached in case test case has finished before
      await waitFor(
        () => this.measures.length * TIME_INTERVAL_IN_MS > duration,
        {
          checkInterval: TIME_INTERVAL_IN_MS,
          timeout: duration * 1000,
        }
      );
      this.measures = this.measures.slice(
        0,
        duration / TIME_INTERVAL_IN_MS + 1
      );
    } else {
      this.shouldStop = true;
      // Hack to wait for the last measures to be received
      await new Promise((resolve) =>
        setTimeout(resolve, TIME_INTERVAL_IN_MS * 2)
      );
    }

    // Ensure polling has stopped
    this.polling?.stop();

    return {
      time: time ?? 0,
      measures: this.measures,
    };
  }
}
