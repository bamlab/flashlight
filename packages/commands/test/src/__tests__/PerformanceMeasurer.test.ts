import "../utils/test/mockChildProcess";
import { LogLevel, Logger } from "@perf-profiler/logger";
import { PerformanceMeasurer } from "../PerformanceMeasurer";
import { emitMeasure, perfProfilerMock } from "../utils/test/mockEmitMeasures";

Logger.setLogLevel(LogLevel.SILENT);

const loggerDebug = jest.spyOn(Logger, "debug");
const loggerError = jest.spyOn(Logger, "error");

describe("PerformanceMeasurer", () => {
  it("handles c++ errors correctly", async () => {
    const measurer = new PerformanceMeasurer("com.example", {
      recordOptions: {
        record: false,
      },
    });
    await measurer.start();
    emitMeasure(0);
    emitMeasure(1);
    perfProfilerMock.stderr?.emit("data", "CPP_ERROR_CANNOT_OPEN_FILE /proc/1234/tasks/578/stat");
    emitMeasure(2);

    expect(loggerDebug).toHaveBeenCalledWith(
      "CPP_ERROR_CANNOT_OPEN_FILE /proc/1234/tasks/578/stat"
    );
    expect(loggerError).not.toHaveBeenCalled();

    expect(measurer.measures).toHaveLength(2);
    expect(measurer.measures).toMatchSnapshot();

    // Reset measures when pid changes
    perfProfilerMock.stderr?.emit("data", "CPP_ERROR_MAIN_PID_CLOSED 1234");
    emitMeasure(0);
    expect(measurer.measures).toHaveLength(0);

    emitMeasure(1);
    emitMeasure(2);

    expect(measurer.measures).toHaveLength(2);
  });
});
