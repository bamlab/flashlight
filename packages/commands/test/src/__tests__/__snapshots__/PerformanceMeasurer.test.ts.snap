// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerformanceMeasurer handles c++ errors correctly 1`] = `
[
  {
    "cpu": {
      "perCore": {
        "0": 12,
        "1": 2,
        "2": 0,
        "3": 50,
        "4": 0,
        "6": 100,
        "7": 2,
      },
      "perName": {
        "ADB-JDWP Connec": 0,
        "AsyncTask #1": 0,
        "BMScavenger": 0,
        "Binder #1": 0,
        "Binder #2": 0,
        "Binder #2 (2)": 0,
        "Binder #3": 0,
        "Binder #4": 0,
        "Binder #5": 0,
        "CookieMonsterBa": 0,
        "CookieMonsterCl": 0,
        "FinalizerDaemon": 0,
        "FinalizerWatchd": 0,
        "FlipperConnecti": 0,
        "FlipperEventBas": 0,
        "FutureTimekeepr": 0,
        "HeapTaskDaemon": 0,
        "HybridData Dest": 0,
        "JITWorker": 0,
        "Jit thread pool": 2,
        "OkHttp Dispatch": 2,
        "OkHttp Dispatch (2)": 4,
        "OkHttp TaskRunn": 0,
        "Profile Saver": 0,
        "ReferenceQueueD": 0,
        "RenderThread": 6,
        "Signal Catcher": 0,
        "alhost:8097/...": 0,
        "com.example": 48,
        "hwuiTask0": 0,
        "hwuiTask1": 0,
        "mali-cmar-backe": 2,
        "mali-mem-purge": 0,
        "mali-utility-wo": 0,
        "mali-utility-wo (2)": 0,
        "mali-utility-wo (3)": 0,
        "mali-utility-wo (4)": 0,
        "mqt_js": 100,
        "mqt_native_modu": 2,
        "perfetto_hprof_": 0,
        "queued-work-loo": 0,
        "queued-work-loo (2)": 0,
        "queued-work-loo (3)": 0,
      },
    },
    "fps": 60,
    "ram": 375.76171875,
    "time": 500,
  },
  {
    "cpu": {
      "perCore": {
        "0": 0,
        "1": 0,
        "2": 0,
        "3": 0,
        "4": 0,
        "6": 0,
        "7": 0,
      },
      "perName": {
        "ADB-JDWP Connec": 0,
        "AsyncTask #1": 0,
        "BMScavenger": 0,
        "Binder #1": 0,
        "Binder #2": 0,
        "Binder #2 (2)": 0,
        "Binder #3": 0,
        "Binder #4": 0,
        "Binder #5": 0,
        "CookieMonsterBa": 0,
        "CookieMonsterCl": 0,
        "FinalizerDaemon": 0,
        "FinalizerWatchd": 0,
        "FlipperConnecti": 0,
        "FlipperEventBas": 0,
        "FutureTimekeepr": 0,
        "HeapTaskDaemon": 0,
        "HybridData Dest": 0,
        "JITWorker": 0,
        "Jit thread pool": 0,
        "OkHttp Dispatch": 0,
        "OkHttp Dispatch (2)": 0,
        "OkHttp TaskRunn": 0,
        "Profile Saver": 0,
        "ReferenceQueueD": 0,
        "RenderThread": 0,
        "Signal Catcher": 0,
        "alhost:8097/...": 0,
        "com.example": 0,
        "hwuiTask0": 0,
        "hwuiTask1": 0,
        "mali-cmar-backe": 0,
        "mali-mem-purge": 0,
        "mali-utility-wo": 0,
        "mali-utility-wo (2)": 0,
        "mali-utility-wo (3)": 0,
        "mali-utility-wo (4)": 0,
        "mqt_js": 0,
        "mqt_native_modu": 0,
        "perfetto_hprof_": 0,
        "queued-work-loo": 0,
        "queued-work-loo (2)": 0,
        "queued-work-loo (3)": 0,
      },
    },
    "fps": 60,
    "ram": 375.76171875,
    "time": 1000,
  },
]
`;
